prov/efa/src/rdm/efa_rdm_rma.c

```diff
@@ -429,27 +429,54 @@ ssize_t efa_rdm_rma_post_write(struct efa_rdm_ep *ep, struct efa_rdm_ope *txe)
 	return efa_rdm_ope_post_send(txe, ctrl_type);
 }
 
-ssize_t efa_rdm_rma_writemsg(struct fid_ep *ep,
+static inline ssize_t efa_rdm_generic_writemsg(struct efa_rdm_ep *efa_rdm_ep,
 			 const struct fi_msg_rma *msg,
 			 uint64_t flags)
 {
 	ssize_t err;
-	struct efa_rdm_peer *peer;
-	struct efa_rdm_ep *efa_rdm_ep;
 	struct efa_rdm_ope *txe;
-	fi_addr_t tmp_addr;
-	struct fi_msg_rma *msg_clone;
-	void *shm_desc[EFA_RDM_IOV_LIMIT];
-	void **tmp_desc;
 	struct util_srx_ctx *srx_ctx;
 
 	EFA_DBG(FI_LOG_EP_DATA,
 	       "write iov_len %lu flags: %lx\n",
 	       ofi_total_iov_len(msg->msg_iov, msg->iov_count),
 	       flags);
 
-	efa_rdm_ep = container_of(ep, struct efa_rdm_ep, base_ep.util_ep.ep_fid.fid);
 	srx_ctx = efa_rdm_ep_get_peer_srx_ctx(efa_rdm_ep);
+	ofi_genlock_lock(srx_ctx->lock);
+
+	txe = efa_rdm_rma_alloc_txe(efa_rdm_ep, msg, ofi_op_write, flags);
+	if (OFI_UNLIKELY(!txe)) {
+		efa_rdm_ep_progress_internal(efa_rdm_ep);
+		err = -FI_EAGAIN;
+		goto out;
+	}
+
+	err = efa_rdm_rma_post_write(efa_rdm_ep, txe);
+	if (OFI_UNLIKELY(err)) {
+		efa_rdm_ep_progress_internal(efa_rdm_ep);
+		efa_rdm_txe_release(txe);
+	}
+out:
+	ofi_genlock_unlock(srx_ctx->lock);
+	return err;
+}
+
+ssize_t efa_rdm_rma_writemsg(struct fid_ep *ep,
+			 const struct fi_msg_rma *msg,
+			 uint64_t flags)
+{
+	ssize_t err;
+	struct efa_rdm_peer *peer;
+	struct efa_rdm_ep *efa_rdm_ep;
+	fi_addr_t tmp_addr;
+	struct fi_msg_rma *msg_clone;
+	void *shm_desc[EFA_RDM_IOV_LIMIT];
+	void **tmp_desc;
+
+	efa_rdm_ep = container_of(ep, struct efa_rdm_ep, base_ep.util_ep.ep_fid.fid);
+	efa_perfset_start(efa_rdm_ep, perf_efa_tx);
```

Comment:
I think this should be inside the generic one, same for the end

Reply:
done